library/core/src/ptr/const_ptr.rs

```diff
@@ -152,18 +152,27 @@ impl<T: ?Sized> *const T {
 
     /// Gets the "address" portion of the pointer.
     ///
-    /// This is equivalent to `self as usize`, which semantically discards
-    /// *provenance* and *address-space* information. To properly restore that information,
-    /// use [`with_addr`][pointer::with_addr] or [`map_addr`][pointer::map_addr].
+    /// This is similar to `self as usize`, which semantically discards *provenance* and
+    /// *address-space* information. However, unlike `self as usize`, it is not possible to cast the
+    /// returned address back to a dereferencable pointer via `as *const T`. To properly restore the
```

Comment:
```suggestion
    /// *address-space* information. However, unlike `self as usize`, casting the returned address
    /// back to a pointer via `as *const T` yields a pointer which is undefined behavior to dereference.
    /// To properly restore the
```
Saying "it is not possible to X" is a bit confusing in this context, since if you actually try it it will almost certainly "work". Probably better to use the words "undefined behavior" directly to clarify the issue.

Reply:
Technically, a dereference of size 0 is actually fine... The fully precise thing to say is that it yields a pointer that is equivalent to `ptr::invalid`. Do you think that would be better?

Reply:
```suggestion
    /// *address-space* information. However, unlike `self as usize`, casting the returned address
    /// back to a pointer yields [`ptr::invalid`][], which is undefined behavior to dereference.
    /// To properly restore the
```
I can't find a way to slip in the ZST stuff without it sounding clunky. Perhaps just mention this on `invalid` docs?